core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/sendsemantics/partitioners/ParallelBatchingPartitioner.scala

```diff
@@ -1,29 +1,70 @@
 package edu.uci.ics.amber.engine.architecture.sendsemantics.partitioners
 
+import edu.uci.ics.amber.engine.common.Constants
 import edu.uci.ics.amber.engine.common.ambermessage.{DataFrame, DataPayload, EndOfUpstream}
 import edu.uci.ics.amber.engine.common.tuple.ITuple
 import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
 
+import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
 abstract class ParallelBatchingPartitioner(batchSize: Int, receivers: Seq[ActorVirtualIdentity])
     extends Partitioner {
-  var batches: Array[Array[ITuple]] = _
-  var currentSizes: Array[Int] = _
 
+  // A bucket corresponds to a partition. When Reshape is not enabled, a bucket has just one receiver.
+  // Reshape divides a skewed partition onto multiple workers. So, with Reshape, a bucket can have
+  // multiple receivers. First receiver in the bucket is the original receiver for that partition.
+  val numBuckets = receivers.length
+  var bucketsToReceivers = new mutable.HashMap[Int, ArrayBuffer[ActorVirtualIdentity]]()
+  var receiverToBatch = new mutable.HashMap[ActorVirtualIdentity, Array[ITuple]]()
+  var receiverToCurrBatchSize = new mutable.HashMap[ActorVirtualIdentity, Int]()
+  val samplingSize =
+    2000 // For every `samplingSize` tuples, record the tuple count to each receiver.
+  var tupleIndexForSampling = 0 // goes from 0 to `samplingSize` and then resets to 0
+  var receiverToWorkloadSamples = new mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]]()
+  @volatile var currentSampleCollectionIndex =
```

Comment:
why is this volatile? are there multiple threads using it?

Reply:
Yes. `addTupleToBatch` (DP thread) uses it to know the index where it has to record samples. `getWorkloadSamples`  (Main thread) resets it to 0. Race condition may be created but it doesn't matter because a tiny count difference in sample doesn't affect anything

Reply:
ok